src/lib/pubkey/ec_group/ec_group.cpp

```diff
@@ -72,97 +65,106 @@ class EC_Group_Data_Map final {
                                                       const BigInt& cofactor,
                                                       const OID& oid,
                                                       EC_Group_Source source) {
+         BOTAN_ASSERT_NOMSG(oid.has_value());
+
          lock_guard_type<mutex_type> lock(m_mutex);
 
          for(auto i : m_registered_curves) {
-            /*
-            * The params may be the same but you are trying to register under a
-            * different OID than the one we are using, so using a different
-            * group, since EC_Group's model assumes a single OID per group.
-            */
-            if(!oid.empty() && !i->oid().empty() && i->oid() != oid) {
-               continue;
-            }
-
-            const bool same_oid = !oid.empty() && i->oid() == oid;
-            const bool same_params = i->params_match(p, a, b, g_x, g_y, order, cofactor);
-
-            /*
-            * If the params and OID are the same then we are done, just return
-            * the already registered curve obj.
-            */
-            if(same_params && same_oid) {
-               return i;
-            }
-
-            /*
-            * If same params and the new OID is empty, then that's ok too
-            */
-            if(same_params && oid.empty()) {
-               return i;
-            }
-
-            /*
-            * Check for someone trying to reuse an already in-use OID
-            */
-            if(same_oid && !same_params) {
-               throw Invalid_Argument("Attempting to register a curve using OID " + oid.to_string() +
-                                      " but a distinct curve is already registered using that OID");
+            if(i->oid() == oid) {
+               /*
+               * If both OID and params are the same then we are done, just return
+               * the already registered curve obj.
+               *
+               * If the params don't match then the user attempted to create some
+               * other set of parameters under an already known OID
+               */
+               if(i->params_match(p, a, b, g_x, g_y, order, cofactor)) {
+                  return i;
+               } else {
+                  throw Invalid_Argument("Attempting to register a curve using OID " + oid.to_string() +
+                                         " but a distinct curve is already registered using that OID");
+               }
```

Comment:
Proper nit: Control flow would be a little easier to understand if we sifted out the error case first and then return without the `else`, IMO.

```suggestion
               if(!i->params_match(p, a, b, g_x, g_y, order, cofactor)) {
                  throw Invalid_Argument("Attempting to register a curve using OID " + oid.to_string() +
                                         " but a distinct curve is already registered using that OID");
               }
               
               return i;
```

Reply:
Makes sense, applied